audio/pwaudio.c

```diff
@@ -0,0 +1,776 @@
+	PWVoice v;
+} PWVoiceIn;
+
+static void stream_destroy(void *data)
+{
+	PWVoice *v = (PWVoice *) data;
+	spa_hook_remove(&v->stream_listener);
+	v->stream = NULL;
+}
+
+/*output data processing function to read stuffs from the buffer*/
+static void playback_on_process(void *data)
+{
+	PWVoice *v = (PWVoice *) data;
+	void *p;
+	struct pw_buffer *b;
+	struct spa_buffer *buf;
+	uint32_t n_frames, req, index, n_bytes;
+	int32_t avail;
+
+	/*obtain a buffer to read from */
+	if ((b = pw_stream_dequeue_buffer(v->stream)) == NULL) {
+		pw_log_warn("out of buffers: %m");
+		return;
+	}
+
+	buf = b->buffer;
+	if ((p = buf->datas[0].data) == NULL)
+		return;
+	if ((req = b->requested * v->frame_size) == 0)
+		req = 4096 * v->frame_size;
+	n_frames = SPA_MIN(req, buf->datas[0].maxsize);
+	n_bytes = n_frames * v->frame_size;
+
+	/*get no of available bytes to read data from buffer */
+
+	avail = spa_ringbuffer_get_read_index(&v->ring, &index);
+
+	if (!v->enabled) {
+		avail = 0;
+	}
+
+	if (avail == 0) {
+		memset(p, 0, n_bytes);
+	} else {
+		if (avail < (int32_t) n_bytes)
+			n_bytes = avail;
+
+		spa_ringbuffer_read_data(&v->ring,
+					 v->buffer, RINGBUFFER_SIZE,
+					 index & RINGBUFFER_MASK, p, n_bytes);
+
+		index += n_bytes;
+		spa_ringbuffer_read_update(&v->ring, index);
+	}
+
+	buf->datas[0].chunk->offset = 0;
+	buf->datas[0].chunk->stride = v->frame_size;
+	buf->datas[0].chunk->size = n_bytes;
+
+	/*queue the buffer for playback */
+	pw_stream_queue_buffer(v->stream, b);
+}
+
+/*output data processing function to generate stuffs in the buffer*/
+static void capture_on_process(void *data)
+{
+	PWVoice *v = (PWVoice *) data;
+	void *p;
+	struct pw_buffer *b;
+	struct spa_buffer *buf;
+	int32_t filled;
+	uint32_t index, offs, n_bytes;
+
+	/*obtain a buffer */
+	if ((b = pw_stream_dequeue_buffer(v->stream)) == NULL) {
+		pw_log_warn("out of buffers: %m");
+		return;
+	}
+
+	/*Write data into buffer */
+	buf = b->buffer;
+	if ((p = buf->datas[0].data) == NULL)
+		return;
+	offs = SPA_MIN(buf->datas[0].chunk->offset, buf->datas[0].maxsize);
+	n_bytes =
+	    SPA_MIN(buf->datas[0].chunk->size, buf->datas[0].maxsize - offs);
+
+	filled = spa_ringbuffer_get_write_index(&v->ring, &index);
+
+	if (!v->enabled) {
+		n_bytes = 0;
+	}
+
+	if (filled < 0) {
+		pw_log_warn("%p: underrun write:%u filled:%d",
+			    p, index, filled);
+	} else if ((uint32_t) filled + n_bytes > RINGBUFFER_SIZE) {
+		pw_log_warn("%p: overrun write:%u filled:%d + size:%u > max:%u",
+			    p, index, filled, n_bytes, RINGBUFFER_SIZE);
+	}
+	spa_ringbuffer_write_data(&v->ring,
+				  v->buffer, RINGBUFFER_SIZE,
+				  index & RINGBUFFER_MASK,
+				  SPA_PTROFF(p, offs, void), n_bytes);
+	index += n_bytes;
+	spa_ringbuffer_write_update(&v->ring, index);
+
+	/*queue the buffer for playback */
+	pw_stream_queue_buffer(v->stream, b);
+}
+
+static void on_stream_state_changed(void *_data, enum pw_stream_state old,
+				    enum pw_stream_state state,
+				    const char *error)
+{
```

Comment:
`old` is unused, is it intended?

Reply:
yes, It's intended. when creating events for a stream, it collects these 4 params on when the stream state changes. this "old" represents the old state of the stream and "state" representing the new state of the stream, we only care about the new state of the stream. but old would be needed when parsing values to .state_changed defined in stream.h